model/message_broker.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/recovery"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+const topicsCollection = "topics"
+
+// MessageEntry represents an entry for a "message" of a topic in the database.
+type MessageEntry struct {
+	// Topic is the name of the topic.
+	Topic string `bson:"topic"`
+	// ForeignID is the id of a document located in another collection used
+	// by the Topic interface to extract the actual message. The id of each
+	// message entry is an int sequentially incremented for each message
+	// published to the topic.
+	ForeignID string `bson:"foreign_id"`
+
+	// ResumeToken and Error should be checked when watching a topic.
+	ResumeToken []byte `bson:"-"`
+	Err         error  `bson:"-"`
+}
+
+var (
+	messageEntryTopicKey     = bsonutil.MustHaveTag(MessageEntry{}, "Topic")
+	messageEntryForeignIDKey = bsonutil.MustHaveTag(MessageEntry{}, "ForeignID")
+)
+
+// PublishToTopic creates a new message entry with the given id in the
+// corresponding collection for the given topic. If the topic does not exist in
+// the global registry, an error is returned.
+func PublishToTopic(ctx context.Context, env cedar.Environment, name string, id string) error {
+	topic, err := GetTopic(name)
+	if err != nil {
+		return err
+	}
+
+	return publishToTopic(ctx, env, topic, id)
+}
+
+// WatchTopic returns a MessageEntry channel fed by a change stream on the
+// topic's collection. The resume token specifies where to start after with the
+// change stream. If the topic does not exist in the global registry, an error
+// is returned.
+func WatchTopic(ctx context.Context, env cedar.Environment, name string, resumeToken []byte) (chan MessageEntry, error) {
+	topic, err := GetTopic(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return watchTopic(ctx, env, topic, resumeToken)
+}
+
+// publishToTopic creates a new message entry with the foreign ID in the
+// corresponding collection for the given topic.
+func publishToTopic(ctx context.Context, env cedar.Environment, topic Topic, foreignID string) error {
+	entry := &MessageEntry{
+		Topic:     topic.Name(),
+		ForeignID: foreignID,
+	}
+	insertResult, err := env.GetDB().Collection(topicsCollection).InsertOne(ctx, entry)
+	grip.DebugWhen(err == nil, message.Fields{
+		"collection":   topic.Name(),
+		"insertResult": insertResult,
+		"op":           "publish to a topic",
+	})
+
+	return errors.Wrapf(err, "problem publishing to topic %s", topic.Name())
+}
+
+func watchTopic(ctx context.Context, env cedar.Environment, topic Topic, resumeToken []byte) (chan MessageEntry, error) {
+	opts := options.ChangeStream()
+	if len(resumeToken) > 0 {
+		opts = opts.SetResumeAfter(resumeToken)
+	}
+	cs, err := env.GetDB().Collection(topicsCollection).Watch(
+		ctx,
+		mongo.Pipeline{{{
+			Key: "$match", Value: bson.D{{
+				Key: messageEntryTopicKey, Value: topic.Name()},
```

Comment:
[the docs imply](https://docs.mongodb.com/manual/changeStreams/#modify-change-stream-output) the key needs to be `"fullDocument." + messageEntryTopicKey`